File: packages/go/converter.go
package playbook

import (
	"encoding/json"
	"errors"
	"fmt"
)

// PlaybookToJSON parses a PLAYBOOK.md markdown string and returns its JSON representation.
// If parsing produces errors, the first error is returned.
func PlaybookToJSON(markdown string) (string, error) {
	result := ParsePlaybook(markdown)
	if result.Definition == nil {
		if len(result.Errors) > 0 {
			return "", fmt.Errorf("parse error: %s", result.Errors[0].Message)
		}
		return "", errors.New("parse error: unknown error")
	}

	data, err := json.MarshalIndent(result.Definition, "", "  ")
	if err != nil {
		return "", fmt.Errorf("json marshal error: %w", err)
	}

	return string(data), nil
}

// JSONToPlaybook converts a JSON representation back to PLAYBOOK.md markdown.
// This is not yet implemented and will return an error.
func JSONToPlaybook(jsonStr string) (string, error) {
	return "", errors.New("JSONToPlaybook is not yet implemented")
}

File: packages/python/src/playbook_md/converter.py
"""
PLAYBOOK.md Converter

Utility functions for converting between PLAYBOOK.md markdown and JSON.
"""

from __future__ import annotations

import json
from dataclasses import asdict

from .parser import parse_playbook


def _strip_none(obj: object) -> object:
    """Recursively remove keys with None values from dicts, matching the
    TypeScript behaviour where undefined fields are omitted from JSON."""
    if isinstance(obj, dict):
        return {k: _strip_none(v) for k, v in obj.items() if v is not None}
    if isinstance(obj, list):
        return [_strip_none(item) for item in obj]
    return obj


def playbook_to_json(markdown: str) -> str:
    """Parse a PLAYBOOK.md string and return the definition as JSON.

    Args:
        markdown: The raw PLAYBOOK.md markdown content.

    Returns:
        A JSON string of the parse result (definition + warnings + errors).

    Raises:
        ValueError: If parsing produces no definition and there are errors.
    """
    result = parse_playbook(markdown)
    raw = asdict(result)
    cleaned = _strip_none(raw)
    return json.dumps(cleaned, indent=2, ensure_ascii=False)


def json_to_playbook(json_str: str) -> str:
    """Convert a JSON playbook definition back to PLAYBOOK.md markdown.

    Args:
        json_str: A JSON string representing a playbook definition.

    Returns:
        A PLAYBOOK.md markdown string.

    Raises:
        NotImplementedError: This function is not yet implemented.
    """
    raise NotImplementedError(
        "json_to_playbook is not yet implemented. "
        "Contributions welcome at https://github.com/PLAYBOOK-MD/playbook-integrations"
    )
